### ISSUE TITLE
System namespace definitions not found in Microsoft.ReactNative.Managed.CodeGen

### ISSUE THREAD

**Issue by BartoszKlonowski** [microsoft/react-native-windows#6788]
## Environment
1. `npx react-native --version`: **4.13.1**
2. `npx react-native info`:
```
System:
    OS: Windows 10 10.0.17763
    CPU: (4) x64 Intel(R) Xeon(R) CPU @ 2.30GHz
    Memory: 12.73 GB / 15.00 GB
  Binaries:
    Node: 12.11.1 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.10 - C:\Program Files\nodejs\yarn.CMD
    npm: 6.11.3 - C:\Program Files\nodejs\npm.CMD
    Watchman: Not Found
  SDKs:
    Android SDK: Not Found
    Windows SDK:
      AllowDevelopmentWithoutDevLicense: Enabled
      AllowAllTrustedApps: Enabled
      Versions: 10.0.10240.0, 10.0.16299.0, 10.0.17134.0, 10.0.17763.0, 10.0.18362.0, 10.0.19041.0
  IDEs:
    Android Studio: Not Found
    Visual Studio: 16.8.30804.86 (Visual Studio Community 2019)
  Languages:
    Java: 12.0.2 - C:\Program Files\OpenJDK\jdk-12.0.2\bin\javac.EXE
    Python: 3.7.3 - C:\tools\miniconda3\python.EXE
  npmPackages:
    @react-native-community/cli: Not Found
    react: 16.13.1 => 16.13.1
    react-native: 0.63.2 => 0.63.2
    react-native-windows: 0.63.10 => 0.63.10
  npmGlobalPackages:
    *react-native*: Not Found
```

## Steps To Reproduce

1. Create new project from the template, just like described [in the official docs](https://microsoft.github.io/react-native-windows/docs/getting-started) or use exisiting project
2. Install packages using `npm install` or `yarn install`
3. build project as release or debug for any architecture:  `npx react-native run-windows --arch x64 --no-deploy --no-launch --no-packager`

## Expected Results
Project is built successfully.

## Actual Results

Build fails with 2932 errors saying that (for example):
>Predefined type 'System.Boolean' is not defined or imported

This error regards each `System.` usage in the *Microsoft.ReactNative.Managed.CodeGen* .

The final error gives a hint that
> error NETSDK1004: Assets file 'C:\Users\circleci\project\node_modules\react-native-windows\Microsoft.ReactNative.Managed.CodeGen\obj\project.assets.json' not found. Run a NuGet package restore to generate this file.

But doing `nuget restore` or `msbuild -t:restore`  leads to another type of failure reported in #6564 
so I believe that this is not a good direction and the problem is in `System.` namespace resolution.

### NOTE:

The same project compiles and runs just fine with the same steps but on a different setup:
```
System:
    OS: Windows 10 10.0.18363
    CPU: (8) x64 Intel(R) Core(TM) i5-10210U CPU @ 1.60GHz
    Memory: 14.55 GB / 23.76 GB
  Binaries:
    Node: 12.9.1 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.5 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 6.10.2 - C:\Program Files\nodejs\npm.CMD
    Watchman: Not Found
  SDKs:
    Android SDK: Not Found
    Windows SDK:
      AllowDevelopmentWithoutDevLicense: Enabled
      AllowAllTrustedApps: Enabled
      Versions: 10.0.16299.0, 10.0.18362.0, 10.0.19041.0
  IDEs:
    Android Studio: Not Found
    Visual Studio: 16.8.30711.63 (Visual Studio Community 2019)
  Languages:
    Java: Not Found
    Python: Not Found
  npmPackages:
    @react-native-community/cli: Not Found
    react: 16.13.1 => 16.13.1
    react-native: 0.63.2 => 0.63.2
    react-native-windows: 0.63.10 => 0.63.10
  npmGlobalPackages:
    *react-native*: Not Found
```

**Comment by asklar** [comment]
@BartoszKlonowski can you make sure you have all the development dependencies installed? http://aka.ms/rnw-deps
If so, can you paste the output of `dotnet info` ?
Also I noticed your version of 0.63 is a little old, try deleting your node_modules and re-running yarn install / npm install to install the latest versions

**Comment by BartoszKlonowski** [comment]
@asklar  Thank you for the quick response. Thing is that this issue happens on the CI server, which means that when the repository is checked out there are no *node_modules* and my steps do include installing them with `yarn`/`npm`.
Is it then really necessary to reinstall them again?

---

As for the development dependencies: Due to the issue occurring on the CI server - I can't say for sure if they are installed, but I've been previously thinking the same and did include the 
```
Set-ExecutionPolicy Unrestricted -Scope Process -Force; iex (New-Object System.Net.WebClient).DownloadString('https://raw.githubusercontent.com/microsoft/react-native-windows/master/vnext/Scripts/rnw-dependencies.ps1')
```
as the first step of my CI config - same error.
So I believe they are installed.

However, I'm pasting the `dotnet --info` output from the CI executor, perhaps this will give us some clues:
```
dotnet --info
.NET SDK (reflecting any global.json):
 Version:   5.0.101
 Commit:    d05174dc5a

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17763
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.101\

Host (useful for support):
  Version: 5.0.1
  Commit:  b02e13abab

.NET SDKs installed:
  2.2.401 [C:\Program Files\dotnet\sdk]
  3.0.100 [C:\Program Files\dotnet\sdk]
  5.0.101 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

**Comment by asklar** [comment]
Once you run yarn, the script can be found under \node_modules\react-native-windows\scripts. You can then run the script with `-NoPrompt` to see what your CI might be missing.
The other thing to do is to upgrade the version of react-native-windows in your package.json. If you're on 63 then 0.63.15 is the latest, please try upgrading.

**Comment by BartoszKlonowski** [comment]
@asklar  Unfortunatelly all requirements are met - I've double checked by running the script, according to your advise:
```
PS C:\Users\circleci\project> .\node_modules\react-native-windows\Scripts\rnw-dependencies.ps1 -NoPrompt
.\node_modules\react-native-windows\Scripts\rnw-dependencies.ps1 -NoPrompt
Checking Free space on C: > 15 GB                            OK
Checking Installed memory >= 16 GB                Failed (warn)
Checking Windows version > 10.0.16299.0                      OK
Checking Developer mode is on                                OK
Checking Long path support is enabled                        OK
Checking Choco                                               OK
Checking git                                                 OK
Checking VS 2019 with UWP and Desktop/C++                    OK
Checking NodeJS 12, 13 or 14 installed                       OK
Checking Chrome                                          Failed
Checking Yarn                                                OK
```

I've also upgraded the react-native-windows version to the latest 0.63.15.

The error remains the same.

**Comment by asklar** [comment]
@BartoszKlonowski thanks for checking. I would imagine the error has to do with how the paths to different .net versions are set up (it's likely that .net core 5 is getting picked up to compile instead of .net core 3). Without a local repro for us to try it out it's going to be hard to pinpoint what exactly is happening. What CI system is this happening in? If you able to reproduce this failure locally, I'd be happy to take a look.

**Comment by BartoszKlonowski** [comment]
@asklar I think I solved the problem!
Or at least I found the workaround... but I'm able to run `npx react-native run-windows` successfully on the CI server.

What I did was perform the Windows solution build "manually" by running following steps:
1. `yarn install`
2. `nuget restore .\windows\Solution.sln`
3. `msbuild -t:restore .\windows\Solution.sln`
4. `msbuild .\windows\Solution.sln`

After these steps, `npx react-native run-windows` builds the app successfully.
So now my CI config file contains these additional steps.

---

However, the question is why do these steps allow to build the RNW app without previous error?
I believe that this must be related to building the *Solution.sln* Windows solution first independently.

**Comment by asklar** [comment]
@BartoszKlonowski glad to hear you're unblocked! 
run-windows will do a `nuget restore` for the solution. In 0.63, this is achieved by calling `nuget.exe`. Afterwards, it calls msbuild.
It's possible that something is messing up and using the wrong version of nuget.exe.
VS 16.8 / NET Core 5 made changes in the interaction between nuget and msbuild, which I think are throwing off our logic. I believe the errors you are getting about project.assets.json are coming from the fact that it's trying to restore the nuget packages for the solution as if the project were .net core 5, which it is not. 

We changed the logic in 0.63 to use the newer nuget 5.8.0 in #6547 but maybe something isn't 100% right there.

@dannyvv and @NickGerleman  FYI.

**Comment by NickGerleman** [comment]
I did a quick scan of the published version of 0.63-stable and far as I can tell we should be using a current NuGet binary. Didn't look much further.

A theory could be that something left a state where the NuGet binary couldn't determine it needed to do a restore (compare to needing to do a force yarn install). This is a pretty flakey justification though, so we should definitely see if more similar errors come in to understand if we have a more common failure mechanism.

**Comment by chrisglein** [comment]
Looks like @BartoszKlonowski is unblocked. Let's keep an eye out to see if anyone else hits something similar, but for now let's consider this resolved.
